File: admin-panel/src/services/payoutsService.ts
import api from '../lib/axios'

export type PayoutRunStatus = 'draft' | 'review' | 'approved' | 'scheduled' | 'processing' | 'paid' | 'failed' | 'cancelled'
export type PayoutRunType = 'driver_weekly' | 'operator_monthly' | 'referral' | 'vendor'
export type PayeeStatus = 'pending' | 'ready' | 'hold' | 'bank_fail' | 'paid' | 'cancelled'
export type AdjustmentType = 'deduction' | 'addition' | 'clawback' | 'penalty'

export interface PayoutRun {
  id: string
  run_ref: string
  run_type: PayoutRunType
  status: PayoutRunStatus
  period_label: string
  period_start: string | null
  period_end: string | null
  payee_count: number
  gross_amount: number
  deduction_amount: number
  hold_amount: number
  net_amount: number
  scheduled_at: string | null
  approved_by: string | null
  approved_at: string | null
  rejection_reason: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface PayoutRunListResponse {
  items: PayoutRun[]
  total: number
  page: number
  page_size: number
}

export interface PayoutAdjustment {
  id: string
  adjustment_type: AdjustmentType
  description: string
  amount: number
  reference: string | null
  created_at: string
}

export interface PayoutPayee {
  id: string
  run_id: string
  entity_type: string
  entity_id: string
  entity_name: string
  entity_ref: string | null
  trip_count: number
  gross_amount: number
  incentive_amount: number
  deduction_amount: number
  hold_amount: number
  net_amount: number
  status: PayeeStatus
  bank_account_ref: string | null
  utr_number: string | null
  paid_at: string | null
  hold_reason: string | null
  created_at: string
  updated_at: string
  adjustments: PayoutAdjustment[]
}

export interface PayoutPayeeListResponse {
  items: PayoutPayee[]
  total: number
  page: number
  page_size: number
}

export interface CreatePayoutRun {
  run_type: PayoutRunType
  period_label: string
  period_start?: string | null
  period_end?: string | null
  notes?: string
}

export const payoutsService = {
  listRuns: (params?: {
    page?: number
    page_size?: number
    status?: string
    run_type?: string
    search?: string
  }) => api.get<PayoutRunListResponse>('/payouts/runs', { params }).then(r => r.data),

  getRun: (id: string) =>
    api.get<PayoutRun>(`/payouts/runs/${id}`).then(r => r.data),

  createRun: (body: CreatePayoutRun) =>
    api.post<PayoutRun>('/payouts/runs', body).then(r => r.data),

  updateRun: (id: string, body: Partial<PayoutRun>) =>
    api.patch<PayoutRun>(`/payouts/runs/${id}`, body).then(r => r.data),

  approveRun: (id: string, notes?: string) =>
    api.post<PayoutRun>(`/payouts/runs/${id}/approve`, { notes }).then(r => r.data),

  rejectRun: (id: string, reason: string) =>
    api.post<PayoutRun>(`/payouts/runs/${id}/reject`, { reason }).then(r => r.data),

  deleteRun: (id: string) =>
    api.delete(`/payouts/runs/${id}`).then(r => r.data),

  listPayees: (runId: string, params?: {
    page?: number
    page_size?: number
    status?: string
    search?: string
  }) => api.get<PayoutPayeeListResponse>(`/payouts/runs/${runId}/payees`, { params }).then(r => r.data),

  getPayee: (payeeId: string) =>
    api.get<PayoutPayee>(`/payouts/payees/${payeeId}`).then(r => r.data),

  updatePayee: (payeeId: string, body: Partial<PayoutPayee>) =>
    api.patch<PayoutPayee>(`/payouts/payees/${payeeId}`, body).then(r => r.data),

  addAdjustment: (payeeId: string, body: {
    adjustment_type: AdjustmentType
    description: string
    amount: number
    reference?: string
  }) => api.post<PayoutAdjustment>(`/payouts/payees/${payeeId}/adjustments`, body).then(r => r.data),
}

File: backend/app/services/payouts_service.py
from __future__ import annotations

from typing import Any, Dict

from fastapi import HTTPException
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.payout import (
    AdjustmentType,
    PayeeStatus,
    PayoutAdjustment,
    PayoutPayee,
    PayoutRun,
    PayoutRunStatus,
)


# ── Payout Runs ────────────────────────────────────────────────────────────────

async def list_runs(
    db: AsyncSession,
    page: int,
    page_size: int,
    status: str | None,
    run_type: str | None,
    search: str | None,
) -> Dict[str, Any]:
    q = select(PayoutRun)
    if status:
        q = q.where(PayoutRun.status == status)
    if run_type:
        q = q.where(PayoutRun.run_type == run_type)
    if search:
        like = f"%{search}%"
        q = q.where(
            PayoutRun.run_ref.ilike(like) | PayoutRun.period_label.ilike(like)
        )
    total_result = await db.execute(select(func.count()).select_from(q.subquery()))
    total = total_result.scalar_one()
    q = q.order_by(PayoutRun.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
    result = await db.execute(q)
    items = result.scalars().all()
    return {"items": items, "total": total, "page": page, "page_size": page_size}


async def get_run(db: AsyncSession, run_id: str) -> PayoutRun:
    result = await db.execute(
        select(PayoutRun).where(PayoutRun.id == run_id)
    )
    run = result.scalar_one_or_none()
    if not run:
        raise HTTPException(status_code=404, detail="Payout run not found")
    return run


async def create_run(db: AsyncSession, data: Dict[str, Any]) -> PayoutRun:
    run = PayoutRun(**data)
    run.status = PayoutRunStatus.draft
    db.add(run)
    await db.commit()
    await db.refresh(run)
    return run


async def update_run(db: AsyncSession, run_id: str, data: Dict[str, Any]) -> PayoutRun:
    run = await get_run(db, run_id)
    for k, v in data.items():
        setattr(run, k, v)
    await db.commit()
    await db.refresh(run)
    return run


async def approve_run(db: AsyncSession, run_id: str, admin_id: str, notes: str | None) -> PayoutRun:
    run = await get_run(db, run_id)
    if run.status not in (PayoutRunStatus.review, PayoutRunStatus.draft):
        raise HTTPException(status_code=400, detail="Run cannot be approved in current status")
    run.status = PayoutRunStatus.approved
    run.approved_by = admin_id
    from datetime import datetime
    run.approved_at = datetime.utcnow()
    if notes:
        run.notes = notes
    await db.commit()
    await db.refresh(run)
    return run


async def reject_run(db: AsyncSession, run_id: str, reason: str) -> PayoutRun:
    run = await get_run(db, run_id)
    run.status = PayoutRunStatus.failed
    run.rejection_reason = reason
    await db.commit()
    await db.refresh(run)
    return run


async def delete_run(db: AsyncSession, run_id: str) -> None:
    run = await get_run(db, run_id)
    if run.status not in (PayoutRunStatus.draft, PayoutRunStatus.failed, PayoutRunStatus.cancelled):
        raise HTTPException(status_code=400, detail="Only draft/failed/cancelled runs can be deleted")
    await db.delete(run)
    await db.commit()


# ── Payout Payees ──────────────────────────────────────────────────────────────

async def list_payees(
    db: AsyncSession,
    run_id: str,
    page: int,
    page_size: int,
    status: str | None,
    search: str | None,
) -> Dict[str, Any]:
    await get_run(db, run_id)
    q = select(PayoutPayee).options(selectinload(PayoutPayee.adjustments)).where(
        PayoutPayee.run_id == run_id
    )
    if status:
        q = q.where(PayoutPayee.status == status)
    if search:
        like = f"%{search}%"
        q = q.where(
            PayoutPayee.entity_name.ilike(like) | PayoutPayee.entity_ref.ilike(like)
        )
    total_result = await db.execute(select(func.count()).select_from(q.subquery()))
    total = total_result.scalar_one()
    q = q.order_by(PayoutPayee.net_amount.desc()).offset((page - 1) * page_size).limit(page_size)
    result = await db.execute(q)
    items = result.scalars().all()
    return {"items": items, "total": total, "page": page, "page_size": page_size}


async def get_payee(db: AsyncSession, payee_id: str) -> PayoutPayee:
    result = await db.execute(
        select(PayoutPayee)
        .options(selectinload(PayoutPayee.adjustments))
        .where(PayoutPayee.id == payee_id)
    )
    payee = result.scalar_one_or_none()
    if not payee:
        raise HTTPException(status_code=404, detail="Payee not found")
    return payee


async def add_payee(db: AsyncSession, run_id: str, data: Dict[str, Any]) -> PayoutPayee:
    await get_run(db, run_id)
    payee = PayoutPayee(run_id=run_id, **data)
    db.add(payee)
    await db.commit()
    await db.refresh(payee)
    return await get_payee(db, payee.id)


async def update_payee(db: AsyncSession, payee_id: str, data: Dict[str, Any]) -> PayoutPayee:
    payee = await get_payee(db, payee_id)
    for k, v in data.items():
        setattr(payee, k, v)
    if data.get("status") == PayeeStatus.paid:
        from datetime import datetime
        payee.paid_at = datetime.utcnow()
    await db.commit()
    await db.refresh(payee)
    return await get_payee(db, payee_id)


async def add_adjustment(
    db: AsyncSession, payee_id: str, data: Dict[str, Any], admin_id: str
) -> PayoutAdjustment:
    await get_payee(db, payee_id)
    adj = PayoutAdjustment(payee_id=payee_id, created_by=admin_id, **data)
    db.add(adj)
    # Recompute net on payee
    payee_result = await db.execute(select(PayoutPayee).where(PayoutPayee.id == payee_id))
    payee = payee_result.scalar_one()
    amount = float(data["amount"])
    if data.get("adjustment_type") in (AdjustmentType.deduction, AdjustmentType.clawback, AdjustmentType.penalty):
        payee.deduction_amount = float(payee.deduction_amount) + abs(amount)
    else:
        payee.incentive_amount = float(payee.incentive_amount) + abs(amount)
    payee.net_amount = (
        float(payee.gross_amount)
        + float(payee.incentive_amount)
        - float(payee.deduction_amount)
        - float(payee.hold_amount)
    )
    await db.commit()
    await db.refresh(adj)
    return adj
